Transparent services

mauriciovasquezbernal opened the pull request:
Transparent cubes differ from standard cubes because they don't have a ports concept, they have only an ingress and egress pipeline and only let the packets go or drop them.

This PR implements this new kind of cube and ports some few services to that approach.

mauriciovasquezbernal commented on src/services/pcn-nat/src/Nat.h:
```diff
@@ -56,19 +55,19 @@ struct sm_v {
 
-class Nat : public polycube::service::Cube<Ports>, public NatInterface {
-  friend class Ports;
+class Nat : public polycube::service::TransparentCube, public NatInterface {
```
It is here: https://github.com/polycube-network/polycube/pull/51/commits/76ab443bee7e7faac9d8e9ffbe1932915d4c51ff#diff-7b40ba7af56da14c5b6e842f2344817aR43

mauriciovasquezbernal replied:
@goldenrye do you have further comments about this PR?

acloudiator commented:
> This PR implements this new kind of cube and ports some few services to that approach.

Adding some explanation from @mauriciovasquezbernal commit description for better reference in PR.
```
Transparent cubes are composed by two sets of eBPF programs, ingress and egress, those
are invoked when there is an incomming/outgoing packet on the nic.

An eBPF program in a transparent cube can pass the packet (modifying it in some
cases like the nat) or drop it.
```
Few questions:
1. Do we've any (simple) flow diagram explaining the transparent cube?
2. Do the Jenkins covers this features test?

acloudiator commented:
> I would suggest to remove `/pcn-ddosmitigator/examples` folder, since outdated and no longer needed.
> 
> As a comment about usability improvements, not to be included in this PR, but for future:
> 
> * How can we simply tell users if a service is transparent or not, via CLI?
> * Not immediately clear to me difference between connect and attach, also an inexpert user can be confused.
> * A little bit lost when `polycubectl ddos1 ?` and I have to come back to `polycubectl ?` and understand there is `attach` primitive. I would suggest to think to an alias for both connect and attach primitives under the service instance. E.g.
> 
> ```
> polycubectl ddos1 ?
> polycubectl ddos1 attach ....
> 
> polycubectl bridge1 ?
> polycubectl bridge1 connect port1 ...
> ```
> In other words, a way to automatically suggest to the user `connect` or `attach` according to service currently used.
> 
> * Since number of commands is increasing, and typos are coming more frequent, an autocompletion support for CLI will increase for sure usability.

@mbertrone  Very valid points from the end-users perspective.
@mauriciovasquezbernal `Attach` and `connect` are really confusing they both synonyms to `join`. Also, we've to mention about it in the document,
https://github.com/polycube-network/polycube/blob/master/Documentation/polycubectl/polycubectl.rst
The above points sound valuable input for future enhancement, but one thing that we might want to see if we can implement in CLI to at least list or identify the transparent service(s). What do you think?

mauriciovasquezbernal commented:
@mbertrone I removed the examples folder of the ddosmitigator and also updated the documentation for it and nat.

It is still missing to update the datamodel for the nat, exposing the external IP as a leaf on the module instead as a parameter of a port.

Another thing we are missing is to update pyang-swagger to generate the code for transparent services, it has problems as it expects the ports to be present. However this work overlaps in some way with what is done in https://github.com/polycube-network/polycube/pull/54 and the future work of improving the developing process, so I'd avoid it now.

About the possibility of having an alias, I think this needs further discussion, I am not sure this is a good idea.

@acloudiator the attach and detach are already documented: https://github.com/polycube-network/polycube/blob/pr/transparent_services/Documentation/cubes.rst#attach-and-detach-primitives and https://github.com/polycube-network/polycube/blob/pr/transparent_services/Documentation/polycubectl/polycubectl.rst.

Currently this is not super easy to implement the possibility to identify if a service is transparent or not, specially because it would conflict with https://github.com/polycube-network/polycube/pull/54, once again, I prefer to implement this feature after that PR is merged.

mbertrone commented:
> @mbertrone I removed the examples folder of the ddosmitigator and also updated the documentation for it and nat.

Thanks!
> 
> It is still missing to update the datamodel for the nat, exposing the external IP as a leaf on the module instead as a parameter of a port.
> 
> Another thing we are missing is to update pyang-swagger to generate the code for transparent services, it has problems as it expects the ports to be present. However this work overlaps in some way with what is done in #54 and the future work of improving the developing process, so I'd avoid it now.

Agree, it should not be part of this PR.
> 
> About the possibility of having an alias, I think this needs further discussion, I am not sure this is a good idea.

For sure it is a good idea to have further discussion later, probably after some feedback from "non internal" people.
> 
> @acloudiator the attach and detach are already documented: https://github.com/polycube-network/polycube/blob/pr/transparent_services/Documentation/cubes.rst#attach-and-detach-primitives and https://github.com/polycube-network/polycube/blob/pr/transparent_services/Documentation/polycubectl/polycubectl.rst.
> 
> Currently this is not super easy to implement the possibility to identify if a service is transparent or not, specially because it would conflict with #54, once again, I prefer to implement this feature after that PR is merged.

Agree.

frisso commented on Documentation/cubes.rst:
```diff
@@ -5,24 +5,67 @@ A ``Cube`` is a logical entity, composed by a ``data plane`` (which can include
+A trasparent cube:
+  - does not define any port, but it *attaches* to existing ports such as the ones of a normal service (e.g., port1 on router2) or a network interface in the host (e.g., veth0);
+  - it *inherits* all the parameters associated to ports (e.g., MAC/IPv4 addresses, link speed, etc.) from the actual port it is attached to (e.g., port 1 on router2, or veth0);
+  - cannot be *connected* to other services, but has to be *attached* to an existing port.
```
==>
(merge first and third item):
- does not define any port, hence cannot be *connected* to other services, but has to be *attached* to an existing port such as the one of a normal service (e.g., port1 on router2) or a network interface in the host (e.g., veth0);

frisso commented on Documentation/developers.rst:
```diff
@@ -252,7 +252,7 @@ If the service is intended to receive data packets in the control path (using an
 +++++++++++++++++++++++++++
 
-The `Port` class contains the `send_packet_out(const std::vector<uint8_t> &packet, Sense sense = Sense::EGRESS)` method that allows to inject packets into the datapath, the sense parameter allows to specify if the packet should be sent (`Sense::EGRESS`) or received through the port (`Sense::INGRESS`).
+The `Port` class contains the `send_packet_out(EthernetII &packet, bool recirculate = false)` method that allows to inject packets into the datapath, the recirculate parameter allows to specify if the packet should be sent out of the port (`recirculate = false`) or received through the port (`recirculate = true`).
```
Not clear to me what "received through the port" means. Does it refer to INGRESS/EGRESS?

mauriciovasquezbernal replied:
No, in this case the packet is recirculated, it simulates that the packet is received in the port instead of being sent out of it.

mauriciovasquezbernal commented on Documentation/services/pcn-ddosmitigator/ddosmitigator.rst:
```diff
@@ -10,34 +10,13 @@ Features
-How to use
-----------
-
-Working modes: STACK vs REDIRECT
```
The working modes are not needed anymore with transparent services.

mauriciovasquezbernal commented on Documentation/services/pcn-ddosmitigator/ddosmitigator.rst:
```diff
@@ -47,8 +26,3 @@ This loads the service as an ``XDP`` program in driver mode, hence discarding pa
-Examples
---------
-
-Examples are available under ``/examples`` folder.
```
For now there's not any plan to include examples.

acloudiator replied:
Issue #68 has been opened, so that we can track and revisit it sometime later.
